compiler/optimizations/optimizeOnClauses.cpp

```diff
@@ -411,30 +533,76 @@ optimizeOnClauses(void) {
   if (fNoOptimizeOnClauses)
     return;
 
+  // If we're not using locks, the extern functions in
+  // atomics are local and safe for fast on.
+  if (0 != strcmp(CHPL_ATOMICS, "locks")) {
+    forv_Vec(ModuleSymbol, module, gModuleSymbols) {
+      if( module->hasFlag(FLAG_ATOMIC_MODULE) ) {
+        Vec<FnSymbol*> moduleFunctions = module->getTopLevelFunctions(true);
+        forv_Vec(FnSymbol, fn, moduleFunctions) {
+          if( fn->hasFlag(FLAG_EXTERN) ) {
+            fn->addFlag(FLAG_FAST_ON_SAFE_EXTERN);
+            fn->addFlag(FLAG_LOCAL_FN);
+          }
+        }
+      }
+    }
+  }
+
   compute_call_sites();
 
   forv_Vec(FnSymbol, fn, gFnSymbols) {
-    if (!fn->hasFlag(FLAG_ON_BLOCK))
-      continue;
-    DEBUG_PRINTF("%p (%s in %s:%d): FLAG_ON_BLOCK (block=%p, id=%d)\n",
-                 fn, fn->cname,
-                 toModuleSymbol(fn->defPoint->parentSymbol)->filename,
-                 fn->linenum(), fn->body, fn->id);
-    DEBUG_PRINTF("\tlength=%d\n", fn->body->length());
     Vec<FnSymbol*> visited;
 
-    if (markFastSafeFn(fn, optimize_on_clause_limit, &visited)) {
-      DEBUG_PRINTF("\t[CANDIDATE FOR FAST FORK]\n");
-      fn->addFlag(FLAG_FAST_ON);
+    int is = markFastSafeFn(fn, optimize_on_clause_limit, &visited);
+
+    bool fastFork = isFast(is);
+    bool removeRmemFences = isLocal(is);
 
-      if (fReportOptimizedOn) {
-        ModuleSymbol *mod = toModuleSymbol(fn->defPoint->parentSymbol);
-        INT_ASSERT(mod);
-        if (developer ||
-            ((mod->modTag != MOD_INTERNAL) && (mod->modTag != MOD_STANDARD))) {
+    if (!fn->hasFlag(FLAG_ON_BLOCK))
+      fastFork = false;
+
+    if (!fn->hasEitherFlag(FLAG_WRAPPER_NEEDS_START_FENCE,
```

Comment:
I think I'd be happy if the logic for this could be stuffed into a `removeUnnecessaryFences()` (or whatever) function with a comment mentioning why this needs to be done.